src/dx/subscribe/subscription.rs

```diff
@@ -0,0 +1,751 @@
+                    manager.register(&handler, cursor);
+                }
+            }
+        }
+    }
+
+    fn unsubscribe(&self) {
+        {
+            let mut is_subscribed_slot = self.is_subscribed.write();
+            if !*is_subscribed_slot {
+                return;
+            }
+            *is_subscribed_slot = false;
+        }
+
+        if let Some(client) = self.client().upgrade().clone() {
+            if let Some(manager) = client.subscription_manager().write().as_mut() {
+                // Mark entities as "not in-use" by subscription.
+                self.entity.decrease_subscriptions_count();
+
+                if let Some((_, handler)) = self.clones.read().iter().next() {
+                    let handler: Weak<dyn EventHandler<T, D> + Send + Sync> = handler.clone();
+                    manager.unregister(&handler);
+                }
+            }
+        }
+    }
+}
+
+impl<T, D> EventHandler<T, D> for Subscription<T, D>
+where
+    T: Transport + Send + Sync + 'static,
+    D: Deserializer + Send + Sync + 'static,
+{
+    fn handle_events(&self, cursor: SubscriptionCursor, events: &[Update]) {
+        if !self.is_subscribed() {
+            return;
+        }
+
+        let filtered_events = self.filtered_events(events);
+
+        let mut cursor_slot = self.cursor.write();
+        if let Some(current_cursor) = cursor_slot.as_ref() {
+            cursor
+                .gt(current_cursor)
+                .then(|| *cursor_slot = Some(cursor));
+        } else {
+            *cursor_slot = Some(cursor);
+        }
+
+        // Go through subscription clones and trigger events for them.
+        self.clones.write().retain(|_, handler| {
+            if let Some(handler) = handler.upgrade().clone() {
+                handler
+                    .event_dispatcher
+                    .handle_events(filtered_events.clone());
+                return true;
+            }
+            false
+        });
+    }
+
+    fn subscription_input(&self, include_inactive: bool) -> SubscriptionInput {
+        if !include_inactive && self.entity.subscriptions_count().eq(&0) {
+            Default::default()
+        }
+
+        self.subscription_input.clone()
+    }
+
+    fn invalidate(&self) {
+        {
+            let mut is_subscribed = self.is_subscribed.write();
+            if !*is_subscribed {
+                return;
+            }
+            *is_subscribed = false;
+        }
+        self.entity.decrease_subscriptions_count();
+
+        // Go through subscription clones and invalidate them.
+        self.clones.write().retain(|_, handler| {
+            if let Some(handler) = handler.upgrade().clone() {
+                handler.event_dispatcher.invalidate();
+                return true;
+            }
+            false
+        });
+    }
+
+    fn id(&self) -> &String {
+        &self.id
+    }
+
+    fn client(&self) -> Weak<PubNubClientInstance<T, D>> {
```

Comment:
Why do we return `Weak` and then `upgrade` it into strong pointer? ;o

Reply:
We upgrade it only at the time when we need to use it. Rest of the time `Weak` pointer is stored which breaks retain cycle (subscriptions list stored in manager inside of `PubNubClientInstance`).

Reply:
Strange, my answer went somewhere. 
I'm trying to move away from strong references to the `PubNubClientInstance`. It will be great for it also return `Arc<Self>` when building in instance. When it will be able to pass components weak reference on itself, there won't be a way to create retain cycle which will require special approach to break (like we have right now with spawned EE). `Subscription` is stored in manager inside `PubNubClientInstance` - this is troublesome and the reason of weak reference.
`Subscription` itself already stores weak reference on it received from the entity for which it has been created. Entities so far store strong references.